### PR TITLE
HCX-195 - Notifications Topic List API

### PR DESCRIPTION
**Pull request by shiva-rakshith**

### CODE REVIEW

File: hcx-apis/src/main/java/org/swasth/hcx/controllers/BaseController.java [hunk 1]
```diff
@@ -235,11 +259,11 @@ protected void loadNotifications() throws IOException {
         notificationList =  (List<Notification>) jsonReader.readValue(resource.getInputStream(), List.class);
     }
 
-    protected boolean isValidNotificaitonId(String notificationId){
+    protected boolean isValidNotificationId(String notificationId){
```
Comment (maheshkumargangula) [inline hcx-apis/src/main/java/org/swasth/hcx/controllers/BaseController.java:262]: BaseController should have only common logic. Why do we have notification-specific logic here?

File: hcx-core/hcx-common/src/main/java/org/swasth/common/dto/Response.java [hunk 2]
```diff
@@ -4,6 +4,7 @@
 import com.fasterxml.jackson.annotation.JsonProperty;
 
 import java.util.HashMap;
+import java.util.LinkedHashMap;
```
Comment (maheshkumargangula) [inline hcx-core/hcx-common/src/main/java/org/swasth/common/dto/Response.java:7]: Where are we using the LinkedHashMap?

File: hcx-core/hcx-common/src/test/java/org/swasth/common/dto/NotificationTest.java [hunk 3]
```diff
@@ -21,43 +22,42 @@ public void testNotificationMockData() {
...
-        assertNull(notification.getEventType());
         assertEquals("Status",notification.getStatus());
-        assertEquals("Id",notification.getId());
-        assertEquals("Name",notification.getName());
-        assertNull(notification.getRecipient());
-        assertNull(notification.getSender());
+        assertEquals("Id",notification.getTopicCode());
+        assertEquals("Name",notification.getTitle());
+        assertNull(notification.getAllowedRecipients());
+        assertNull(notification.getAllowedSenders());
```
Comment (maheshkumargangula) [inline hcx-core/hcx-common/src/test/java/org/swasth/common/dto/NotificationTest.java:56]: Please add assertion at the data level for all the fields.

**Review by maheshkumargangula — CHANGES_REQUESTED** [review]

File: hcx-core/hcx-common/src/main/java/org/swasth/common/dto/Response.java (hunk 2 above)
Comment (shiva-rakshith) [reply hcx-core/hcx-common/src/main/java/org/swasth/common/dto/Response.java]: This is an unused import statement, removed id.

File: hcx-core/hcx-common/src/test/java/org/swasth/common/dto/NotificationTest.java (hunk 3 above)
Comment (shiva-rakshith) [reply hcx-core/hcx-common/src/test/java/org/swasth/common/dto/NotificationTest.java]: updated the test cases.

File: hcx-apis/src/main/java/org/swasth/hcx/controllers/BaseController.java (hunk 1 above)
Comment (shiva-rakshith) [reply hcx-apis/src/main/java/org/swasth/hcx/controllers/BaseController.java]: Moved notification related logic to the notification service.

File: hcx-apis/src/main/java/org/swasth/hcx/controllers/BaseController.java [hunk 4]
```diff
@@ -216,13 +217,36 @@ private List<Subscription> fetchSubscriptions(String participantCode) throws Exc
...
+                Iterator<Map<String, Object>> iterator = filteredNotificationList.iterator();
+                while(iterator.hasNext()){
+                    Map<String, Object> notification = iterator.next();
+                    for(String key: request.getFilters().keySet()) {
+                        if(!notification.get(key).equals(request.getFilters().get(key))){
+                            iterator.remove();
+                            break;
+                        }
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (shiva-rakshith) [reply hcx-apis/src/main/java/org/swasth/hcx/controllers/BaseController.java]: Updated the implementation and moved notification related logic to the notification service.

File: hcx-apis/src/main/java/org/swasth/hcx/controllers/BaseController.java [hunk 5]
```diff
@@ -216,13 +217,36 @@ private List<Subscription> fetchSubscriptions(String participantCode) throws Exc
...
     protected ResponseEntity<Object> getNotifications(Map<String, Object> requestBody) throws Exception {
         Response response = new Response();
-        Request request = null;
+        NotificationListRequest request = null;
         try {
             checkSystemHealth();
-            request = new Request(requestBody);
+            if(!requestBody.containsKey(FILTERS) || MapUtils.isEmpty((Map<String,Object>) requestBody.get(FILTERS))) {
+                throw new ClientException(ErrorCodes.ERR_INVALID_NOTIFICATION_REQ, "Notification filters is missing or empty");
+            }
```
[the remark this answers is not in the document]
Comment (shiva-rakshith) [reply hcx-apis/src/main/java/org/swasth/hcx/controllers/BaseController.java]: removed validation for empty filters.

File: api-gateway/src/main/resources/application.yaml [hunk 6]
```diff
@@ -120,4 +120,4 @@ spring:
             - RewritePath=/v0.7/notification/subscription/list, /v1/notification/subscription/list
             - RewritePath=/v0.7/notification/list, /v1/notification/list
           predicates:
-            - Path=/v0.7/participant/**, /health, /service/health, /v0.7/notification/subscription/list, /v0.7/notification/list
+            - Path=/v0.7/participant/**, /health, /service/health, /v0.7/notification/subscription/list, /v0.7/notification/topic/list
```
[the remark this answers is not in the document]
Comment (shiva-rakshith) [reply api-gateway/src/main/resources/application.yaml]: done

File: hcx-apis/src/main/java/org/swasth/hcx/aspect/HealthCheckAspect.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+@Aspect
+@Component
+public class HealthCheckAspect {
+
+    private static final Logger logger = LoggerFactory.getLogger(HealthCheckAspect.class);
+
+    @Before("execution(* org.swasth.hcx.controllers.v1.*.*(..))")
+    public ResponseEntity<Response> healthCheckBeforeEachAPICall() {
+        if (!HealthCheckManager.allSystemHealthResult) {
+            logger.info("Health check is failed");
```
Comment (maheshkumargangula) [inline hcx-apis/src/main/java/org/swasth/hcx/aspect/HealthCheckAspect.java:27]:
```suggestion
            logger.error("Health check is failed");
```
Also, log what is the failure?

File: hcx-apis/src/main/java/org/swasth/hcx/handlers/EventHandler.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    private String postgresTableName;
+
+    public void processAndSendEvent(String metadataTopic, Request request) throws Exception {
+        String payloadTopic = env.getProperty(KAFKA_TOPIC_PAYLOAD);
+        String key = request.getSenderCode();
+        // TODO: check and remove writing payload event to kafka
+        String payloadEvent = eventGenerator.generatePayloadEvent(request);
+        String metadataEvent = eventGenerator.generateMetadataEvent(request);
+        String query = String.format("INSERT INTO %s (mid,data,action,status,retrycount,lastupdatedon) VALUES ('%s','%s','%s','%s',%d,%d)", postgresTableName, request.getMid(), JSONUtils.serialize(request.getPayload()), request.getApiAction(), QUEUED_STATUS, 0, System.currentTimeMillis());
+        System.out.println("Mid: " + request.getMid() + " :: Event: " + metadataEvent);
```
Comment (maheshkumargangula) [inline hcx-apis/src/main/java/org/swasth/hcx/handlers/EventHandler.java:45]:
```suggestion
        logger.debug("Mid: " + request.getMid() + " :: Event: " + metadataEvent);
```

File: hcx-apis/src/main/java/org/swasth/hcx/service/NotificationService.java [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+    @Value("${postgres.subscription.tablename}")
+    private String postgresSubscription;
+
+    @Value("${postgres.subscription.insertQuery}")
+    private String insertSubscription;
+
+    @Value("${postgres.subscription.subscriptionQuery}")
+    private String selectSubscription;
+
+    @Value("${notification.path:classpath:Notifications.json}")
```
Comment (maheshkumargangula) [inline hcx-apis/src/main/java/org/swasth/hcx/service/NotificationService.java:38]:
```suggestion
    @Value("${notification.path:classpath:notifications.json}")
```

File: hcx-apis/src/main/java/org/swasth/hcx/service/NotificationService.java [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+    }
+
+    private void isValidNotificationId(String notificationId) throws ClientException {
+        boolean isValid = false;
+        for (Map<String, Object> notification: notificationList) {
+            if(notificationId.equals(notification.get(TOPIC_CODE)))
+                isValid = true;
+        }
+        if(!isValid) throw new ClientException(ErrorCodes.ERR_INVALID_NOTIFICATION_TOPIC_CODE, "Invalid NotificationId." + notificationId + " is not present in the master list of notifications");
+    }
```
Comment (maheshkumargangula) [inline hcx-apis/src/main/java/org/swasth/hcx/service/NotificationService.java:108]: Why don't you create the topic code list in init?

File: hcx-apis/src/test/resources/application-test.yml [hunk 11]
```diff
@@ -58,6 +55,7 @@ hcx:
 
 #Notifications related configurations
 notification:
+  path: ${notification_path:classpath:Notifications.json}
```
Comment (maheshkumargangula) [inline hcx-apis/src/test/resources/application-test.yml:58]:
```suggestion
  path: ${notification_path:classpath:notifications.json}
```

File: hcx-core/hcx-common/src/main/java/org/swasth/common/dto/NotificationListRequest.java [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+import static org.swasth.common.utils.Constants.FILTERS;
+
+public class NotificationListRequest {
+
+    private Map<String,Object> filters;
+
+    public NotificationListRequest(Map<String,Object> requestBody) throws ClientException {
+        if (!requestBody.containsKey(FILTERS))
+            throw new ClientException(ErrorCodes.ERR_INVALID_NOTIFICATION_REQ, "Notification filters property is missing");
+        this.filters = (Map<String, Object>) requestBody.get(Constants.FILTERS);
```
Comment (maheshkumargangula) [inline hcx-core/hcx-common/src/main/java/org/swasth/common/dto/NotificationListRequest.java:18]: It is fine not to have filters in the request. Consider user in this case applies no filters.

**Review by maheshkumargangula — CHANGES_REQUESTED** [review]

File: hcx-core/hcx-common/src/main/java/org/swasth/common/dto/NotificationListRequest.java (hunk 12 above)
Comment (shiva-rakshith) [reply hcx-core/hcx-common/src/main/java/org/swasth/common/dto/NotificationListRequest.java]: Remove validation.

File: hcx-apis/src/test/resources/application-test.yml (hunk 11 above)
Comment (shiva-rakshith) [reply hcx-apis/src/test/resources/application-test.yml]: updated the filename

File: hcx-apis/src/main/java/org/swasth/hcx/service/NotificationService.java (hunk 10 above)
Comment (shiva-rakshith) [reply hcx-apis/src/main/java/org/swasth/hcx/service/NotificationService.java]: updated the implementation

File: hcx-apis/src/main/java/org/swasth/hcx/service/NotificationService.java (hunk 9 above)
Comment (shiva-rakshith) [reply hcx-apis/src/main/java/org/swasth/hcx/service/NotificationService.java]: updated the filename

File: hcx-apis/src/main/java/org/swasth/hcx/handlers/EventHandler.java (hunk 8 above)
Comment (shiva-rakshith) [reply hcx-apis/src/main/java/org/swasth/hcx/handlers/EventHandler.java]: added logger

File: hcx-apis/src/main/java/org/swasth/hcx/aspect/HealthCheckAspect.java (hunk 7 above)
Comment (shiva-rakshith) [reply hcx-apis/src/main/java/org/swasth/hcx/aspect/HealthCheckAspect.java]: updated the logger

**Review by maheshkumargangula — APPROVED** [review]

**Pull request merged**